### DISCUSSION TITLE
Need help with send data to Tuya device (when writing a converter)

### DISCUSSION THREAD

**Discussion by prokudin07** [Koenkk/zigbee2mqtt#14866] [Devices]
Hello. 
my experience:  wrote few converters for some devices in old description for tuya devices
there is:
 Tuya din relay from this [issue](https://github.com/Koenkk/zigbee2mqtt/issues/13097) and [discussion](https://github.com/Koenkk/zigbee2mqtt/discussions/13663)

Made: 
almost wrote whole converter with new description for Tuya devices

**Code in /device/tuya.js**
```
    {
        fingerprint: tuya.fingerprint('TS0601', ['_TZE200_lsanae15']),
        model: 'TS0601_din_1',
        vendor: 'TuYa',
        description: 'Zigbee DIN energy meter',
        fromZigbee: [tuya.fzDataPoints],
        toZigbee: [tuya.tzDataPoints],
        configure: tuya.configureMagicPacket,
        exposes: [
                  tuya.exposes.switch(),
                  e.energy(),
                  e.power(),
                  e.voltage(),
                  e.current(),
                  exposes.enum('fault',ea.STATE,['clear', 'over current threshold', 'over power threshold', 'over voltage threshold','wrong frequency threshold>
                        .withDescription('Text of fault'),
                  exposes.text('threshold_raw',ea.STATE),
                  exposes.binary('threshold1_state',ea.STATE,'on','off')
                        .withDescription('if ON - relay will be off when threshold reached'),
                  exposes.text('threshold1_description', ea.STATE),
                  exposes.text('threshold1_value', ea.STATE)
                        .withDescription('Setup value on the device'),
                  exposes.text('threshold2_state',ea.STATE)
                        .withDescription('if ON - relay will be off when threshold reached'),
                  exposes.text('threshold2_description', ea.STATE),
                  exposes.text('threshold2_value', ea.STATE)
                        .withDescription('Setup value on the device'),
                  exposes.binary('clear_event', ea.STATE_SET, 'ON', 'OFF')
                        .withDescription('Clear event'),
                  exposes.text('meterid',ea.STATE).withDescription('Meter ID')
                 ],
        meta: {
            tuyaDatapoints: [
                [1, 'energy', tuya.valueConverter.divideBy100],
                [3, null, null], //Monthly, but sends data only after request. for user z2m - think no matter
                [4, null, null], //Dayly, but sends data only after request. for user z2m - think no matter
                [6, null, tuya.valueConverter.phaseB], // voltage and current
                [10, 'fault', tuya.valueConverterBasic.lookup({'clear': 0, 'over current threshold': 1, 'over power threshold': 2, 'over voltage threshold': 4,>
                [11, null, null], //Frozen - strange function, in native app - nothing is clear
                [16, 'state', tuya.valueConverter.onOff],
                [17, null, tuya.valueConverter.alarm],
                [18, 'meterid', tuya.valueConverter.raw],
                [20, 'clear_event', tuya.valueConverter.onOff], //Clear event
                [21, null, null], //Forward Energy T1 - don't know what this
                [22, null, null], //Forward Energy T2 - don't know what this
                [23, null, null], //Forward Energy T3 - don't know what this
                [24, null, null], //Forward Energy T4 - don't know what this
            ],
        },
        whiteLabel: [{vendor: 'Hiking', model: 'DDS238-2'}, {vendor: 'TuYa', model: 'RC-MCB'}],
    },
```
**Code in lib/tuya.js**
```
   alarm: {
        from: (v) => {
            const buffer = Buffer.from(v, 'base64');
            const lookup = {
                0: 'not set',
                1: 'Over current threshold',
                3: 'Over voltage threshold',
           };
            const lookup_state = {0: 'off', 1: 'on'};
            return {
                threshold_raw: buffer,
                threshold1_state: lookup_state[buffer[1]],
                threshold1_description: lookup[buffer[0]],
                threshold1_value : (buffer[3] | buffer[2] << 8),
                threshold2_state: lookup_state[buffer[5]],
                threshold2_description: lookup[buffer[4]],
                threshold2_value : (buffer[7] | buffer[6] << 8),
            };
        },
    },
```
 **threshold_raw** DP17 - expose with originally raw data from device after decode 'base64' 
see on screenshot. (1) - raw data, (2)- fields and values
![photo_2022-11-07 22 24 54](https://user-images.githubusercontent.com/81035252/200400595-5ed54526-3e04-4f94-bccb-0dcfb694d9ae.jpeg)

**Question 1:**
Give me direction  for DP 17 - i want to do  this field for writing.
Understand  it's in  block:
```
        to: (v) => {


        },
```
but how code to base64 - don't know how.

**Question 2:**

 **threshold_raw** DP17 - get value only when device first join to z2m and then no changes data  when I changed that on device.
How can I forcibly get  from device?

Thanks in advance for replies

P.S. I know how to work with "ea.STATE_SET".  :-)

**Comment by Koenkk** [comment]
Is dp 17 settable when this device is used via the TuYa gateway?

**Reply by prokudin07** [reply]
Yes, settable. 
![photo_2022-11-08 20 35 20](https://user-images.githubusercontent.com/81035252/200635702-7214cce5-7b09-49cc-9194-4ab845431b67.jpeg)
![photo_2022-11-08 20 35 15](https://user-images.githubusercontent.com/81035252/200635738-95a901b1-ad12-4d1b-bc3b-97f8361eb52a.jpeg)
![photo_2022-11-08 20 35 24](https://user-images.githubusercontent.com/81035252/200635760-dbd17687-63ed-4d19-a651-11b1d1ab2565.jpeg)

And maybe set up threshold_state to: null, 0 or 1.  
a) null - threshold not set up
b) 0 - threshold   set up, but relay will not to off when threshold reached
c) 1 - threshold   set up and relay will to off when threshold reached
And possible 3 situations:
1. threshold1_state and threshold2_state - null
2. one of both - null
3. both - no null

**Reply by Koenkk** [reply]
Then the to code should be something like this:

```js
const buffer = Buffer.from([0x00, 0x01]); // 0x00, 0x01 should be the payload, basically the reversed thing of the "from" part
return buffer.toString('base64');
```

**Reply by prokudin07** [reply]
Thanks, but I can not do converter for DP17 ((

**Reply by Koenkk** [reply]
What appears in the "State" tab of the device when using the following?

```
[17, null, tuya.valueConverter.raw],
```

**Reply by prokudin07** [reply]
I wrote next code - [17, null, tuya.valueConverter.alarm]
```
    alarm: {
        from: (v) => {
            const buffer = Buffer.from(v, 'base64');
            const lookup = {
                0: 'not set',
                1: 'Over current threshold',
                3: 'Over voltage threshold',
           };
            const lookup_state = {0: 'off', 1: 'on'};
            return {
                threshold_raw: buffer,
                threshold1_state: lookup_state[buffer[1]],
                threshold1_description: lookup[buffer[0]],
                threshold1_value : (buffer[3] | buffer[2] << 8),
                threshold2_state: lookup_state[buffer[5]],
                threshold2_description: lookup[buffer[4]],
                threshold2_value : (buffer[7] | buffer[6] << 8),
            };
        },
    },
```
After convert value in that place
 **const buffer = Buffer.from(v, 'base64');**
Got next value
 buffer = {"0":1,"1":0,"2":0,"3":80,"4":3,"5":1,"6":0,"7":245} - if both threshold setup
or
buffer= {"0":1,"1":0,"2":0,"3":80} - if only one threshold setup
where 
threshold1_state  - OFF (0) - just alarm, ON (1) - relay will be off when threshold reached
threshold1_description - text 
 threshold1_value - in example - 80 Ampere when threshold reached

I've tried  change  only one value, an example for "1":0 to "1"":1 
then that string buffer = {"0":1,"1":**1**,"2":0,"3":80,"4":3,"5":1,"6":0,"7":245}
convert to base64 (try also without convert and convert to JSON)   and then send to device in block 
```
            to: (v) => {
                
                return v;
            },
```
of course changed  description expose "ea.STATE_SET" 

but got error or nothing in that [place](https://github.com/Koenkk/zigbee-herdsman-converters/blob/15e67d9428a0c143b5bd91d791c657e525cb13bd/lib/tuya.js#L1437)

**Comment by Fleudo** [comment]
Friends, I bought a meter Hiking DTS 238-7 Wifi see IVAP and would like to get ways to have access to all censors. For example: I want to set up a daily table of total positive energy produced, total inverse energy produced, total energy injected into the grid. Does anyone know the device and would dare to make a code for the Home Assistant?
